[SPARK-6980] [CORE] Akka timeout exceptions indicate which conf controls them (RPC Layer)

BryanCutler opened the pull request:
Latest changes after refactoring to the RPC layer.  I rebased against trunk to make sure to get any recent changes since it had been a while.  I wasn't crazy about the name `ConfigureTimeout` and `RpcTimeout` seemed to fit better, but I'm open to suggestions!

I ran most of the tests and they pass, but others would get stuck with "WARN TaskSchedulerImpl: Initial job has not accepted any resources".  I think its just my machine, so I'd though I would push what I have anyway.  

Still left to do:
- I only added a couple unit tests so far, there are probably some more cases to test
- Make sure all uses require a `RpcTimeout`
- Right now, both the `ask` and `Await.result` use the same timeout, should we differentiate between these in the TimeoutException message?
- I wrapped `Await.result` in `RpcTimeout`, should we also wrap `Await.ready`?
- Proper scoping of classes and methods

@hardmettle, feel free to help out with any of these!

BryanCutler committed:
[SPARK-6980] Akka ask timeout description refactored to RPC layer

BryanCutler committed:
[SPARK-6980] Using RpcTimeout.awaitResult for future in AppClient now

BryanCutler committed:
[SPARK-6980] Added some RpcTimeout unit tests

BryanCutler committed:
[SPARK-6980] Minor cleanup and scala style fix

AmplabJenkins commented:
Can one of the admins verify this patch?

squito commented:
ok to test

squito commented on core/src/main/scala/org/apache/spark/util/RpcUtils.scala:
```diff
@@ -47,14 +46,24 @@ object RpcUtils {
+    }
+    catch {
+      case _: Throwable =>
+        RpcTimeout(conf, "spark.network.timeout", "120s")
```
Can you change this to catch a much narrower exception?

Or even better -- how about you move this into `RpcTimeout.apply` -- have it take a prioritized list of properties to check, and it will check the conf for each of them, avoiding any need for exception handling

BryanCutler replied:
I like the second idea, I'll add that in another commit

squito commented on core/src/main/scala/org/apache/spark/util/RpcUtils.scala:
```diff
@@ -47,14 +46,24 @@ object RpcUtils {
+    }
+    catch {
+      case _: Throwable =>
+        RpcTimeout(conf, "spark.network.timeout", "120s")
```
same here

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -94,7 +97,7 @@ private[spark] abstract class RpcEnv(conf: SparkConf) {
    */
   def setupEndpointRefByURI(uri: String): RpcEndpointRef = {
-    Await.result(asyncSetupEndpointRefByURI(uri), defaultLookupTimeout)
+    Await.result(asyncSetupEndpointRefByURI(uri), defaultLookupTimeout.duration)
```
any reason not to use `timeout.awaitResult` here, so we get the better exception msg?

BryanCutler replied:
Nope - I must have missed that one, it's fixed now.

squito commented on core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala:
```diff
@@ -212,7 +211,7 @@ private[spark] class AkkaRpcEnv private[akka] (
   override def asyncSetupEndpointRefByURI(uri: String): Future[RpcEndpointRef] = {
     import actorSystem.dispatcher
-    actorSystem.actorSelection(uri).resolveOne(defaultLookupTimeout).
+    actorSystem.actorSelection(uri).resolveOne(defaultLookupTimeout.duration).
```
though this doesn't follow the same pattern of `Await.result`, can we catch the timeout here too?

hardmettle replied:
If the return type changes from `Future[RpcEndpointRef]` to `RpcEndpointRef` then only the application of `Await.result` is possible else it will not be possible.

BryanCutler replied:
The timeout is a little trickier for Futures.  From what I understand, creating the future is non-blocking, so we can't just call `Await.result`.  If the Future isn't fulfilled within the timeout, its marked as failed, then sometime later if the future is acted upon, the TimeoutException is thrown.

I think we might be able to use the `andThen` method described here: http://doc.akka.io/docs/akka/snapshot/scala/futures.html.  We would need to add a function to `RpcTimeout` that takes in a `Future[T]`, then applies `andThen` to it which checks if the future completed with failure, then returns a `Future[T]`.

squito replied:
what you are describing sounds reasonable.  We certainly don't want to change the return type or await on the result here.

You could potentially fix the doubly-appended msg by throwing a subclass of `TimeoutException`, which then you could check for before you amend a msg.  However, this is really begging the question -- if there is one timeout passed in here, and another is passed to an `Await.result` later, how is akka using them both?  I suppose it will timeout on whichever one is shorter?  I don't see anything very definitive in the docs, perhaps we should confirm on akka-user.  That would inform how we should do the error handling.

(Honestly I thiink it will also be fine to not stress too much about this case, it may not be worth it.)

BryanCutler replied:
I'm going to do some research on this and see what our options are while still keeping the same return type.

squito commented on core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala:
```diff
@@ -293,9 +292,9 @@ private[akka] class AkkaRpcEndpointRef(
+  override def ask[T: ClassTag](message: Any, timeout: RpcTimeout): Future[T] = {
     import scala.concurrent.ExecutionContext.Implicits.global
-    actorRef.ask(AkkaMessage(message, true))(timeout).flatMap {
+    actorRef.ask(AkkaMessage(message, true))(timeout.duration).flatMap {
```
same thing here about catching the timeout

hardmettle commented:
@BryanCutler Do I have the access to merge your PR ?

BryanCutler committed:
[SPARK-6980] Fixed await result not being handled by RpcTimeout

hardmettle commented:
@BryanCutler can I get write access to your PR repo ? I want to merge my changes , how do I do it ?

BryanCutler commented:
@hardmettle are you able to checkout my branch, then maybe you could share any changes/additions as a pull-request?

BryanCutler committed:
[SPARK-6980] Added creation of RpcTimeout with Seq of property keys

BryanCutler commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,89 @@ private[spark] object RpcAddress {
+   * @param timeoutPropList prioritized list of property keys for the timeout in seconds
+   * @param defaultValue default timeout value in seconds if no properties found
+   */
+  def apply(conf: SparkConf, timeoutPropList: Seq[String], defaultValue: String): RpcTimeout = {
```
Creating `RpcTimeout` from a prioritized list of property keys, from @squito previous request

BryanCutler commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,89 @@ private[spark] object RpcAddress {
+
+    // Find the first set property or use the default value with the first property
+    val foundProp = timeoutPropList.view.map(x => (x, conf.getOption(x))).filter(_._2.isDefined).
+      map(y => (y._1, y._2.get)).headOption.getOrElse(timeoutPropList.head, defaultValue)
```
This might look a little confusing, but it will lazily evaluate the Seq of properties to check if they are set, to get the first one.  If no properties are found, the default value is used.  Have to use a tuple of `(<property-key>, <property-value>)` so that when a property is found, we can use that string in the timeout message.  If uses the default value, the first property in the list will be shown in the timeout message.

squito replied:
I would avoid views, they might work in this case but they have some issues.  (eg. see https://groups.google.com/forum/#!searchin/scala-user/views/scala-user/LG-8y3EaOt0/16Q3MPg4ZEAJ)  You should be able to to do the same thing with `timeoutPropList.iterator`.  But, I actually think this would be more straightforward with a simple while loop over the iterator, I think the chained operations get pretty confusing for the reader.

hardmettle replied:
@BryanCutler created a PR and added your changes .Also I have made the changes suggested by @squito for this line note. I will be adding more test cases . Here it is , please take a look and suggest if any changes are needed . https://github.com/apache/spark/pull/6290

hardmettle replied:
@BryanCutler Sorry for the confusion . I have now corrected it . Created a PR in your branch

Commit:
Retrieving properties from property list using iterator and while loop instead of chained functions

squito commented:
@hardmettle I don't think there are any occurrences of `Await.ready` that are relevant here -- where they do occur, they aren't tied to a configuration variable, the wait is infinite, or its in test code, etc.

Commit:
Changing the loop condition to halt at the first match in the property list for RpcEnv exception catch

Commit:
Modifying loop condition to find property match

BryanCutler committed:
Merge pull request #2 from hardmettle/configTimeoutUpdates_6980

[SPARK-6980] [CORE] [WIP] Creating wrapper for Akka timeout exceptions to get better information using conf (RPC Layer)

squito commented:
@BryanCutler @hardmettle just checking if there is any update here, let me know when its ready to take another look.  Seemed like it was getting close, only "big" thing was what to do with `actorSelection.resolveOne`.

hardmettle commented:
@squito I have been banging my head on this . any suggestions for improvements?

BryanCutler commented:
Hey guys,  I've been playing around with making the futures fail before calling await, here is a test case that does it:
```scala
test("Future failure with RpcTimeout") {

    class EchoActor extends Actor {
      def receive: Receive = {
        case msg =>
          Thread.sleep(50)
          sender() ! msg
      }
    }

    val system = ActorSystem("EchoSystem")
    val echoActor = system.actorOf(Props(new EchoActor), name = "echoA")

    val timeout = new RpcTimeout(50 millis, "spark.rpc.short.timeout")

    val fut = echoActor.ask("hello")(10 millis).mapTo[String].recover {
      case te: TimeoutException => throw timeout.amend(te)
    }

    fut.onFailure {
      case te: TimeoutException => println("failed with timeout exception")
    }

    fut.onComplete {
      case Success(str) => println("future success")
      case Failure(ex) => println("future failure")
    }

    println("sleeping")
    Thread.sleep(50)
    println("Future complete: " + fut.isCompleted.toString() + ", " + fut.value.toString())

    println("Caught TimeoutException: " +
      intercept[TimeoutException] {
        //timeout.awaitResult(fut)  // will print RpcTimeout description twice
        Await.result(fut, 10 millis)
      }.getMessage()
    )
}
```

If we use `recover`, then when a future fails we can check for a `TimeoutException` and then re-raise the exception with our modified message as before, and we wouldn't need to do it in `Await.result` anymore.  It also doesn't change the type of `Future` returned.  I kind of think this is a better solution because it would handle all cases of `Future` timeouts regardless if `Await.result` is used.  What do you guys think?

squito commented:
@BryanCutler haven't looked at this closely yet, but quick question -- are you talking using this just for `actorSelction.resolveOne`?  Or are you proposing doing this in place of what you already have?

BryanCutler commented:
@squito I was thinking to use this in place of what we have, but actually it wouldn't cover the case where `Await.result` times out while the future is still not completed.  To cover all cases we could add a function in `RpcTimeout` that would take a `Future[T]` as input and use `recover` (as in the code above) to  handle a timeout on the future.  Then we would have to subclass `TimeoutException`, as you suggested, so we don't end up amending the message twice.  If you want, I could code this up to see how it would look?

One thing I'm not too sure of though, when using `recover` it requires an `ExecutionContext`.  In the test code I posted, it was fine to import the implicit conversion to the default global context, but I don't know if the same holds true for the rest of Spark.  Any thoughts on this?

squito commented:
Hi @BryanCutler,
I finally took a closer look at this.  I hadn't realized until you brought it up that even `ask` takes one timeout on the original call, and then another one in `Await.result`.  If the timeout to the original `ask` expires, you get an `AskTimeoutException`, eg. `akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://EchoSystem/user/echoA#-553815755]] after [50 ms]`, but if its the timeout in `Await.result`, you just get a `TimeoutException`, eg. `java.util.concurrent.TimeoutException: Futures timed out after [50 milliseconds]`.  Eg:

```
val fut = echoActor.ask("hello")(50 millis)
val inBetweenWait = (15 millis)
Thread.sleep(inBetweenWait.toMillis)
val result = Await.result(fut, 50 millis)
```

for me the behavior seems flip flop somewhere around `inBetweenWait = 15 millis`, but that will most likely vary.  Even if the `ask` timeout is shorter than the `Await` timeout, I still might get the exception from `Await`.

In any case, the reason I'm bringing this up is that we're sort of in the same scenario with `actorSelection.resolveOne` (looks like that timeout is just [passed on to `ask`](https://github.com/akka/akka/blob/1dac4010990e6a0790dedceece83b1e5864c37e7/akka-actor/src/main/scala/akka/actor/ActorSelection.scala#L60)).

Your suggestion of using `recover` works for handling that exception from `ask` or `resolveOne` etc, basically to handle a timeout that might already be baked into a `Future`.  I'd like to keep around what you already have, since  (a) we still need something else for `Await.result` b/c that might timeout first (as you've noted) and (b) its a lot simpler to understand in my opinion.

I think its worth adding your suggestion as well for those remaining cases -- its a nice trick :).  I didn't see any existing uses for those methods, but most likely the resulting `Future`s will just get wrapped in an `Await.result` with the same timeout, but it would still be nice to have that in place.  I think just (a) throwing your own subclass of `TimeoutException` to avoid double appending and (b) having `timeout.addMessageIfTimeout(future)` (or whatever you want to call it) take in an `ExecutionContext` seems fine.  (You can see that `scala.concurrent.ExecutionContext.Implicits.global` is imported elsewhere in `AkkaRpcEnv`.)

phew that was a mouthful.  hope this all makes sense

BryanCutler commented:
Hi @squito,
Yeah that all makes sense and I think we are on the same page, so let me put what we've been talking about in another commit and we can take another look.  I noticed too that the timeout in ask is throwing an `AskTimeoutException`, do think it is ok to be changing it to just a `TimeoutException`?  I didn't find anywhere explicitly catching `AskTimeoutException`, so it might not be a problem.  Thanks!

squito commented:
yeah i think its totally fine for you to throw a `TimeoutException` instead of an `AskTimeoutException`.

Also it would be good to bring this up to date w/ master, and then we can pull in some more reviewers.  thanks

BryanCutler committed:
Merge branch 'master' into configTimeout-6980-tmp

Conflicts:
	core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala
	core/src/main/scala/org/apache/spark/storage/BlockManagerMaster.scala

BryanCutler committed:
[SPARK-6980] Resolved conflicts after master merge

BryanCutler committed:
[SPARK-6980] Added addMessageIfTimeout for when a Future is completed with TimeoutException

BryanCutler commented on core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala:
```diff
@@ -214,8 +213,10 @@ private[spark] class AkkaRpcEnv private[akka] (
+    defaultLookupTimeout.addMessageIfTimeout(
+      actorSystem.actorSelection(uri).resolveOne(_).
+        map(new AkkaRpcEndpointRef(defaultAddress, _, conf))
+    )
```
I was trying to avoid calling `addMessageIfTimeout` and also still having to reference the `timeout.duration`.  Otherwise, the call would look like this
```scala
defaultLookupTimeout.addMessageIfTimeout(
  actorSystem.actorSelection(uri).resolveOne( defaultLookupTimeout.duration ).
    map(new AkkaRpcEndpointRef(defaultAddress, _, conf))
)
```
but as it is, it does look kind of confusing, so I'm not sure what is better

squito replied:
yeah I think its better to be more explicit, its not clear what the first `_` is.  Also I think it would be a bit clearer in two statements, so you can add a comment explaining, eg.

```
val future = actorSystem.actorSelection(uri).resolveOne(defaultLookupTimeout.duration)
  .map(new AkkaRpcEndpointRef(defaultAddress, _, conf))
// this is just in case there is a timeout from creating the future in resolveOne, we want the
// exception to indicate the conf that determines the timeout
defaultLookupTimeout.addMessageIfTimeout(future)
```

BryanCutler commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,130 @@ private[spark] object RpcAddress {
+      // Add a warning message if Future is passed to addMessageIfTimeoutTest more than once
+      case rte: RpcTimeoutException => throw new RpcTimeoutException(rte.getMessage() +
+        " (Future has multiple calls to RpcTimeout.addMessageIfTimeoutTest)")
+      // Any other TimeoutException get converted to a RpcTimeoutException with modified message
```
I noticed that it's ends up pretty easy to call this function multiple times on the same future, and it basically keeps chaining together `recover` callbacks.  This won't cause an error, but maybe not ideal, so I though adding a warning might help.  Actually, as I'm typing this I realized that this warning message could get added multiple times too, so maybe it's best to just take this out.  Uggh, haha!

squito replied:
we could do the same trick with the exceptions, to avoid appending the message multiple times to the message.  You could return a new subclass of `Future`, which just delegates all methods to whatever it wrapped, and this method would be a no-op if called on that subclass.

But IMO, this is just not worth the complication ... honestly I am already slightly regretting taking us down this path, the benefit is pretty marginal.  At least I've learned something along the way, I hope you feel the same :)

squito replied:
(btw, I'm *not* saying we need to get rid of this, just saying I don't think we need to stress about the possibility of multiple messages ...)

BryanCutler commented:
@squito @hardmettle I updated with latest changes from the master branch and added `addMessageIfTimeout` to work with timeouts on `Futures` and a `RpcTimeoutException` class so not to add the message twice.  I still need to add some more unit tests to cover all the cases.

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,130 @@ private[spark] object RpcAddress {
+ * An exception thrown if RpcTimeout modifies a [[TimeoutException]].
+ */
+private[rpc] class RpcTimeoutException(message: String)
+  extends TimeoutException(message)
```
I just realized that `TimeoutException` doesn't allow you to specify the original cause.  This is a little unfortunate, because it means all the stack traces will end where we create this class.  It looks like there is an `initCause` method which does what we want though.  So I think the constructor should require a `cause: TimeoutException`, which we set via `initCause`.

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,130 @@ private[spark] object RpcAddress {
+
+  /** Applies the duration to create future before calling addMessageIfTimeout*/
+  def addMessageIfTimeout[T](f: FiniteDuration => Future[T]): Future[T] = {
+    addMessageIfTimeout(f(duration))
```
given the discussion on the usage, IMO we should get rid of this

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,130 @@ private[spark] object RpcAddress {
+
+    // Find the first set property or use the default value with the first property
+    val itr = timeoutPropList.iterator
+    var foundProp =  None: Option[(String, String)]
```
nit: the usual style for this is `var foundProp: Option[(String, String)] = None`

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,130 @@ private[spark] object RpcAddress {
+      val propKey = itr.next()
+      conf.getOption(propKey) match {
+        case Some(prop) => foundProp = Some(propKey,prop)
+        case None =>
```
this could be `conf.getOption(propKey).foreach { prop => foundProp = Some(propKey, prop) }`.  Its fine as is, though

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,130 @@ private[spark] object RpcAddress {
+  def message: String = description
+
+  /** Amends the standard message of TimeoutException to include the description */
+  def createRpcTimeoutException(te: TimeoutException): RpcTimeoutException = {
```
`private`

squito commented:
looking good @BryanCutler !  left some minor comments.  When you update the tests, also we can add more reviewers

BryanCutler committed:
[SPARK-6980] Changed addMessageIfTimeout to PartialFunction, cleanup from PR comments

BryanCutler commented:
Thanks for the feedback @squito!  I updated the code with fixes from your comments.  I wasn't too happy with the `addMessageIfTimeout` function - it seemed a little too intrusive and confusing for what it was doing.  I changed it slightly to propose that `RpcTimeout` just offers a `PartialFunction` that can be used in a `Future.recover` callback.  I think this way is a little cleaner, but let me know if you prefer the function to use `Futures`, as before, and I can change it back.

BryanCutler commented on core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala:
```diff
@@ -214,8 +213,11 @@ private[spark] class AkkaRpcEnv private[akka] (
+      map(new AkkaRpcEndpointRef(defaultAddress, _, conf)).
+      // this is just in case there is a timeout from creating the future in resolveOne, we want the
+      // exception to indicate the conf that determines the timeout
+      recover(defaultLookupTimeout.addMessageIfTimeout)
```
Here is the usage with `addMessageIfTimeout` as a `PartialFunction`

BryanCutler committed:
[SPARK-6980] Added UT for addMessageIfTimeout, cleaned up UTs

BryanCutler commented:
Hey @squito , I cleaned up the UTs, currently we are testing for:
* askWithRetry timeout TimeoutException has added message
* constructing RpcTimeout from single property
* constructing RpcTimeout from prioritized list of properties
* trying to create a RpcTimeout from non-existing property
* ask pattern successfully completes within timeout
* timeout on Future throws RpcTimeoutException with the added message
* if RpcTimeoutException is already thrown, the message won't be added a second time

Are there any cases I left out?

BryanCutler committed:
[SPARK-6980] Cleaned up UT imports

BryanCutler commented on core/src/test/scala/org/apache/spark/rpc/akka/AkkaRpcEnvSuite.scala:
```diff
@@ -47,4 +56,60 @@ class AkkaRpcEnvSuite extends RpcEnvSuite {
+      // once the future is complete
+      val msg1 =
+        intercept[RpcTimeoutException] {
+          Await.result(fut, 200 millis)
```
I noticed this test sometimes outputs a message that there is a dead letter from the ask message that times out.  Is that something we need to fix?  I looked but couldn't find a simple way to stop the dead letter logging.

squito replied:
when creating your `ActorSystem` above, can you also pass in a config with `log-dead-letters=off`? see http://doc.akka.io/docs/akka/2.3.11/general/configuration.html

squito replied:
I just tried this, it seems to work:

```
    val akkaConf = ConfigFactory.empty().withValue("akka.log-dead-letters",
      ConfigValueFactory.fromAnyRef("off"))
    val system = ActorSystem("EchoSystem", akkaConf)
```

BryanCutler replied:
Great, that seemed to do the trick, Thanks!

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -18,8 +18,11 @@
-import scala.concurrent.{Await, Future}
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.duration._
+import scala.concurrent.{Awaitable, Await, Future}
```
nit: you don't need to import `FiniteDuration` explicitly since you're importing `duration._`.  Also you are supposed to order direct class imports before package imports (not just alphabetically), so it should be:

```
import scala.concurrent.{Awaitable, Await, Future}
import scala.concurrent.duration._
```

the intellij import organizer will get this wrong, but aaron davidson wrote a plugin which does it right -- there are instructions for using it here: https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Imports

BryanCutler replied:
Thanks, using the plugin now

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,117 @@ private[spark] object RpcAddress {
+  }
+
+  /**
+   * Waits for a completed result to catch and amend a TimeoutException message
```
I think it would be a little clearer as:
"Wait for the completed result and return it.  If the result is not available within this timeout, throw a [[TimeoutException]] which indicates which configuration controls the timeout."

BryanCutler replied:
done

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,117 @@ private[spark] object RpcAddress {
+/**
+ * Create an RpcTimeout using a configuration property that controls the timeout duration so when
+ * a TimeoutException is thrown, the property key will be indicated in the message.
+ */
```
I think this comment can be deleted, it doesn't really apply to `object RpcTimeout`, really its for the individual `apply` methods which you've already got good docs for.

BryanCutler replied:
done

squito commented on core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala:
```diff
@@ -295,19 +297,20 @@ private[akka] class AkkaRpcEndpointRef(
-    }(ThreadUtils.sameThread).mapTo[T]
+  override def ask[T: ClassTag](message: Any, timeout: RpcTimeout): Future[T] = {
+    actorRef.ask(AkkaMessage(message, true))(timeout.duration).flatMap {
+        // The function will run in the calling thread, so it should be short and never block.
```
nit: is this block over-indented?  seems like ti shouldn't be changed at all from before

BryanCutler replied:
good catch, it was off.  fixed.

squito commented on core/src/main/scala/org/apache/spark/util/AkkaUtils.scala:
```diff
@@ -17,9 +17,9 @@
 
 package org.apache.spark.util
 
+import org.apache.spark.rpc.RpcTimeout
```
nit: ordering

BryanCutler replied:
done

squito commented on core/src/main/scala/org/apache/spark/util/RpcUtils.scala:
```diff
@@ -17,11 +17,11 @@
 
-import scala.concurrent.duration._
 import scala.language.postfixOps
+import scala.concurrent.duration._
```
nit: ordering (it was right before)

BryanCutler replied:
done

squito commented on core/src/main/scala/org/apache/spark/util/RpcUtils.scala:
```diff
@@ -47,14 +47,14 @@ object RpcUtils {
-      conf.get("spark.network.timeout", "120s")) seconds
+  def askTimeout(conf: SparkConf): RpcTimeout = {
+    RpcTimeout(conf, Seq("spark.rpc.askTimeout",
+      "spark.network.timeout"), "120s")
```
nit: I think this will now fit on one line

BryanCutler replied:
done

squito commented on core/src/main/scala/org/apache/spark/util/RpcUtils.scala:
```diff
@@ -47,14 +47,14 @@ object RpcUtils {
-      conf.get("spark.network.timeout", "120s")) seconds
+  def lookupTimeout(conf: SparkConf): RpcTimeout = {
+    RpcTimeout(conf, Seq("spark.rpc.lookupTimeout",
+      "spark.network.timeout"), "120s")
```
same here, one line

BryanCutler replied:
done

squito commented on core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala:
```diff
@@ -162,9 +163,15 @@ abstract class RpcEnvSuite extends SparkFunSuite with BeforeAndAfterAll {
+        case te: TimeoutException =>
+          assert(te.getMessage().contains(shortProp))
+        case e: Exception =>
+          assert(e.getCause().getMessage().contains(shortProp))
```
Do you know how we end up with a non-`TimeoutException` here?
(I realize that before there was already a check for that, you are just keeping what was there -- just curious if you know)

BryanCutler replied:
The call to `askWithRetry` will catch all exceptions and then throw a `SparkException` with the `TimeoutException` as the cause.  So its the `TimeoutException` case that will never happen, but I left that in there to keep it as before.

squito replied:
ah, I see, that makes sense.  in that case, I kinda prefer we just `intercept[SparkException]` in the first place, and then check the cause & the cause's msg.  And maybe add a brief comment that `askWithRetry` always wraps with `SparkException` so we look at the cause to see the original exception.

BryanCutler replied:
Yeah, that sounds better to me too.  Should I go ahead and make this change?

squito replied:
Sorry @BryanCutler I just noticed your last set of comments and changes -- yes please do make this change.

squito commented on core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala:
```diff
@@ -539,6 +546,30 @@ abstract class RpcEnvSuite extends SparkFunSuite with BeforeAndAfterAll {
+    val rt2 = RpcTimeout(conf, Seq("spark.ask.invalid.timeout", testProp, secondaryProp), "1s")
+    assert( testDurationSeconds === rt2.duration.toSeconds )
+
+    // Try to construct RpcTimeout with an unconfigured property
```
add a case for using the default value: `RPCTimeout(conf, Seq("spark.ask.invalid.timeout"), "1s")`

BryanCutler replied:
done

squito commented on core/src/test/scala/org/apache/spark/rpc/akka/AkkaRpcEnvSuite.scala:
```diff
@@ -17,9 +17,18 @@
 import org.apache.spark.rpc._
 import org.apache.spark.{SecurityManager, SparkConf}
 
+
```
nit: delete extra newline

BryanCutler replied:
done

squito commented on core/src/test/scala/org/apache/spark/rpc/akka/AkkaRpcEnvSuite.scala:
```diff
@@ -47,4 +56,60 @@ class AkkaRpcEnvSuite extends RpcEnvSuite {
+      val msg2 =
+        intercept[RpcTimeoutException] {
+          timeout.awaitResult(fut)
+        }.getMessage()
```
I see what you're doing here, to test both the `recover(defaultLookupTimeout.addMessageIfTimeout)` and also `timeout.awaitResult` -- but can you another case where you just call `timeout.awaitResult` without first doing `Await.result(...)`?

BryanCutler replied:
I was thinking that RpcEnvSuite timeout test would cover that case, although it is wrapped in `askWithRetry` [here] (https://github.com/BryanCutler/spark/blob/configTimeout-6980/core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala#L166).  It would probably be good to add it here too.

squito replied:
yeah you're right, I hadn't thought about that case when I mentioned this ... but another test doesn't hurt :)

BryanCutler replied:
done

squito commented:
Jenkins, ok to test

squito commented:
Jenkins, test this please

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +185,117 @@ private[spark] object RpcAddress {
+ */
+object RpcTimeout {
+
+  private[this] val messagePrefix = "This timeout is controlled by "
```
instead of putting `messagePrefix` in each of the `apply` methods, what if its just used by `RpcTimeout` itself when building its description?  Then the second arg to the `RpctTimeout` constructor could just be the property, not the full description.

The reason I'm asking is because in the tests, where you directly construct an `RpcTimeout`, the `messagePrefix` is missing, so if you look at the exception its kinda weird.  And I think that the real msgs are missing a "." before the message prefix, so it would be nice to look at in the test.

BryanCutler replied:
That's a good point.  Way back when I first made the unit test for `AkkaUtils` the config was being read using `conf.getTimeAsSeconds` like [here](https://github.com/BryanCutler/spark/blob/configTimeout-6980/core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala#L257) and if I tried to configure the tests in the millisecond range, it wouldn't work.  Because of that I couldn't create a RpcTimeout with `apply` and didn't get the `messagePrefix`.  

I don't think this is the case anymore and maybe we should use `conf.getTimeAsMs` to read the property values?  I think it would then automatically convert seconds to milliseconds, and I could fix the tests to use this instead and get the proper message.

squito commented:
hi @BryanCutler this is looking good!  all of my comments are minor things.  When we see if the tests pass, I'll rope in more reviewers.  (Sorry I should have enabled the tests long ago, my mistake)

squito commented:
also you can remove [wip] from the title now I think

BryanCutler committed:
[SPARK-6980] Cleaned up import ordering, comments, spacing from PR feedback

BryanCutler committed:
[SPARK-6980] Added akkaConf to prevent dead letter logging

BryanCutler committed:
[SPARK-6980] Added UT for constructing RpcTimeout with default value

squito commented:
retest this please

squito commented:
please jenkins?

BryanCutler committed:
[SPARK-6980] Cleaned up import ordering

BryanCutler committed:
[SPARK-6980] Added unit test for ask then immediat awaitReply

BryanCutler commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +184,114 @@ private[spark] object RpcAddress {
+
+object RpcTimeout {
+
+  private[this] val messagePrefix = "This timeout is controlled by "
```
That's a good point about the message in the tests.  Way back when I first made the unit test for `AkkaUtils` the config was being read using `conf.getTimeAsSeconds` like [here](https://github.com/BryanCutler/spark/blob/configTimeout-6980/core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala#L257) and if I tried to configure the tests in the millisecond range, it wouldn't work.  Because of that I couldn't create a RpcTimeout with `apply` and didn't get the `messagePrefix`.  

I don't think this is the case anymore and maybe we should use `conf.getTimeAsMs` to read the property values?  I think it would then automatically convert seconds to milliseconds, and I could fix the tests to use this instead and get the proper message.

squito replied:
I don't think this is related to whether or not we get the time as seconds or millis.  I'm thinking that we still leave the `FiniteDuration` in the constructor, but just change that second param to the constructor to be just the name of the conf, not the full description.  eg.  :

```scala
diff --git a/core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala b/core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala
index 5149cf1..e0de397 100644
--- a/core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala
+++ b/core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala
@@ -197,19 +197,16 @@ private[rpc] class RpcTimeoutException(message: String, cause: TimeoutException)
  * Associates a timeout with a description so that a when a TimeoutException occurs, additional
  * context about the timeout can be amended to the exception message.
  * @param timeout timeout duration in seconds
- * @param description description to be displayed in a timeout exception
+ * @param conf the configuration parameter that controls this timeout
  */
-private[spark] class RpcTimeout(timeout: FiniteDuration, description: String) {
+private[spark] class RpcTimeout(timeout: FiniteDuration, val conf: String) {

   /** Get the timeout duration */
   def duration: FiniteDuration = timeout

-  /** Get the message associated with this timeout */
-  def message: String = description
-
   /** Amends the standard message of TimeoutException to include the description */
   private def createRpcTimeoutException(te: TimeoutException): RpcTimeoutException = {
-    new RpcTimeoutException(te.getMessage() + " " + description, te)
+    new RpcTimeoutException(te.getMessage() + ". This timeout is controlled by " + conf, te)
   }
```

(and corresponding changes to the `apply` methods.)

I do kinda wish that the values were read w/ `conf.getTimeAsMs` to allow finer grained resolution, but (a) I'd be nervous about changing any default and (b) in any case that should just be a separate issue.

squito replied:
in fact we could also mark the constructor as `private[rpc]` to be clear that normal use should really go through the `RpcTimeout.apply` methods.

BryanCutler replied:
I see what you're saying.  I kept the `messagePrefix` separated so that if there was ever a need to create a `RpcTimeout` without a conf, it would still be possible and the message could be fit accordingly.  But not a big deal, I'll put in the above patch.  Right now it's always tied to a conf, so it does make sense.

Quick Scala question:  would it be better to make the constructor use `val timeout: FiniteDuration` and remove that getter method too?

BryanCutler replied:
done

BryanCutler commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +184,114 @@ private[spark] object RpcAddress {
+   * @param defaultValue default timeout value in seconds if property not found
+   */
+  def apply(conf: SparkConf, timeoutProp: String, defaultValue: String): RpcTimeout = {
+    val timeout = { conf.getTimeAsSeconds(timeoutProp, defaultValue) seconds }
```
Should this be `conf.getTimeAsMs` instead?

squito replied:
as mentioned above, I don't think so -- lets stick w/ the current behavior of using `conf.getTimeAsSeconds` for now.

squito commented:
Something seems to be wrong w/ jenkins, hopefully it will test it soon.  In any case, @zsxwing could you take a look?  I think the only things left are minor style issues which bryan and I can keep working on, but  I think its ready for to review.

squito commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +184,114 @@ private[spark] object RpcAddress {
+}
+
+
+object RpcTimeout {
```
`private[spark]`

BryanCutler replied:
done

BryanCutler committed:
[SPARK-6980] RpcTimeout object scope should be private[spark]

BryanCutler committed:
[SPARK-6980] Moved MessagePrefix to createRpcTimeoutException directly

BryanCutler commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +184,109 @@ private[spark] object RpcAddress {
+private[spark] class RpcTimeout(timeout: FiniteDuration, val conf: String) {
+
+  /** Get the timeout duration */
+  def duration: FiniteDuration = timeout
```
Is this necessary or should we just use `val duration: FiniteDuration` in the constructor like `val conf: String`?

squito replied:
yeah I think you should just change the constructor

BryanCutler replied:
Sure, that sounds good.  One other minor question @squito , the constructor has the parameter `conf: String` while `apply` has `conf: SparkConf` and `timeoutProp: String` which seems a little unclear.  I think it might be better to stay consistent and have the constructor say `timeoutProp` instead of `conf`, like this
```scala
class RpcTimeout(val duration: FiniteDuration, val timeoutProp: String)
```
what do you think?

squito replied:
yes, I agree that would be more consistent, good point

squito commented:
ping @zsxwing

SparkQA commented:
[Test build #952 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/952/consoleFull) for   PR 6205 at commit [`1394de6`](https://github.com/apache/spark/commit/1394de6afa8040b545ecdb16ead3b76a0597858f).

SparkQA commented:
[Test build #952 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/952/console) for   PR 6205 at commit [`1394de6`](https://github.com/apache/spark/commit/1394de6afa8040b545ecdb16ead3b76a0597858f).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

BryanCutler committed:
[SPARK-6980] Changed UT ask message timeout to explicitly intercept a SparkException

squito commented:
jenkins, retest this please

SparkQA commented:
[Test build #953 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/953/consoleFull) for   PR 6205 at commit [`1394de6`](https://github.com/apache/spark/commit/1394de6afa8040b545ecdb16ead3b76a0597858f).

SparkQA commented:
[Test build #953 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/953/console) for   PR 6205 at commit [`1394de6`](https://github.com/apache/spark/commit/1394de6afa8040b545ecdb16ead3b76a0597858f).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

squito commented:
shoot, those mima failures are real ... looks like `RpcUtils` is public (probably accidentally) so we can't change the return type of `lookupTimeout` and `askTimeout`.  @BryanCutler can you change to new method names that return `RpcTimeout`, and put the old ones back but mark them as deprecated?

BryanCutler committed:
[SPARK-6980] Changed constructor to use val duration instead of getter function, changed name of string property from conf to timeoutProp for consistency

BryanCutler committed:
[SPARK-6980] Had to increase timeout on positive test case because a processor slowdown could trigger an Future TimeoutException

zsxwing commented:
Oops... Forgot to make `RpcUtils` private.

zsxwing commented on core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala:
```diff
@@ -182,3 +184,109 @@ private[spark] object RpcAddress {
+      val propKey = itr.next()
+      conf.getOption(propKey).foreach { prop => foundProp = Some(propKey, prop) }
+    }
+    val finalProp = foundProp.getOrElse(timeoutPropList.head, defaultValue)
```
The above lines can be replaced by a single line: `val finalProp = timeoutPropList.flatMap(conf.getOption).headOption.getOrElse(defaultValue)`.

zsxwing replied:
Never mind. Looks it's long too.

tedyu replied:
If timeoutPropList.head doesn't appear in conf, defaultValue would be returned.
Is this intended ?

zsxwing commented on core/src/test/scala/org/apache/spark/rpc/akka/AkkaRpcEnvSuite.scala:
```diff
@@ -47,4 +57,71 @@ class AkkaRpcEnvSuite extends RpcEnvSuite {
 
+  test("timeout on ask Future with RpcTimeout") {
+
+    class EchoActor(sleepDuration: Long) extends Actor {
```
Is it possible to write this test using the new Rpc API? Then we don't need to rewrite this test when removing Akka.

BryanCutler replied:
Yeah, I'll give it a shot

zsxwing commented on core/src/test/scala/org/apache/spark/rpc/akka/AkkaRpcEnvSuite.scala:
```diff
@@ -17,8 +17,18 @@
+
+import akka.actor.{Actor, ActorSystem, Props}
+import akka.pattern.ask
+
```
nit: extra empty line

zsxwing commented on core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala:
```diff
@@ -539,6 +546,37 @@ abstract class RpcEnvSuite extends SparkFunSuite with BeforeAndAfterAll {
+    assert( rt3.conf.contains(defaultProp) )
+
+    // Try to construct RpcTimeout with an unconfigured property
+    intercept[Throwable] {
```
Could you change `Throwable` to `NoSuchElementException`?

zsxwing commented on core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala:
```diff
@@ -539,6 +546,37 @@ abstract class RpcEnvSuite extends SparkFunSuite with BeforeAndAfterAll {
+    val testDurationSeconds = 30
+    val secondaryProp = "spark.ask.secondary.timeout"
+
+    conf.set(testProp, testDurationSeconds.toString + "s")
```
nit: testDurationSeconds.toString + "s" -> s"${testDurationSeconds}s"

BryanCutler committed:
[SPARK-6980] Put back in deprecated RpcUtils askTimeout and lookupTimout to fix MiMa errors

BryanCutler committed:
[SPARK-6980] Corrected issues from feedback

BryanCutler committed:
[SPARK-6980] Corrected import organization

BryanCutler commented:
Thanks for the feedback @zsxwing and @squito.  I added back the old methods of `askTimeout` and `lookupTimeout` in `RpcUtils` and marked as deprecated.  Running the mima test locally, I still get errors though - but I'm not sure I'm running it correctly.  Is there anywhere else besides the `RpcUtils` changes that might be causing incompatibility?

zsxwing commented:
retest this please

BryanCutler committed:
Merge branch 'master' into configTimeout-6980

BryanCutler commented:
After I merged with the master branch, the MiMa checks pass locally.  I think this commit from yesterday fixed it - [link] (https://github.com/apache/spark/commit/13ae806b255cfb2bd5470b599a95c87a2cd5e978)
cc @squito @zsxwing

BryanCutler commented:
Hey @squito and @zsxwing , I just came across a weird problem... After running spark shell, I noticed this exception being thrown
```
15/06/24 21:37:01 ERROR Utils: Uncaught exception in thread heartbeat-receiver-event-loop-thread
org.apache.spark.SparkException: Error sending message [message = (BlockManagerHeartbeat(BlockManagerId(driver, localhost, 51109)),600 seconds)]
	at org.apache.spark.rpc.RpcEndpointRef.askWithRetry(RpcEndpointRef.scala:118)
	at org.apache.spark.rpc.RpcEndpointRef.askWithRetry(RpcEndpointRef.scala:77)
	at org.apache.spark.scheduler.DAGScheduler.executorHeartbeatReceived(DAGScheduler.scala:188)
	at org.apache.spark.scheduler.TaskSchedulerImpl.executorHeartbeatReceived(TaskSchedulerImpl.scala:371)
	at org.apache.spark.HeartbeatReceiver$$anonfun$receiveAndReply$1$$anon$2$$anonfun$run$2.apply$mcV$sp(HeartbeatReceiver.scala:107)
	at org.apache.spark.util.Utils$.tryLogNonFatalError(Utils.scala:1264)
	at org.apache.spark.HeartbeatReceiver$$anonfun$receiveAndReply$1$$anon$2.run(HeartbeatReceiver.scala:106)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.spark.SparkException: Unmatched message (BlockManagerHeartbeat(BlockManagerId(driver, localhost, 51109)),600 seconds) from Actor[akka://sparkDriver/temp/$To]
```
I traced it down to `BlockManagerHeartbeat` message [here](https://github.com/BryanCutler/spark/blob/configTimeout-6980/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L189)
I couldn't figure out why the message was not matched, and how it was even compiling it.. but actually it looks like the message type is being made into a tuple of the original message and the `FiniteDuration` - nice, haha!  I checked and didn't see this happening anywhere else, and once I fixed this up the problem went away.  So my questions are

* Is this something that we need to prevent from accidentally happening again?
* To fix this line, can we put the 600 seconds in a conf property?  The other option for a `RpcTimeout` would look like this
```scala
BlockManagerHeartbeat(blockManagerId), new RpcTimeout(600 seconds, "BlockManagerHeartbeat"))
```

zsxwing commented:
> Is this something that we need to prevent from accidentally happening again?

I encountered this issue. Scala compiler will try to pack all parameters into a Tuple when it cannot find a match method. Maybe we should consider adding the compiler flag `-Ywarn-adapted-args` to make the compiler output a warning, or `-Yno-adapted-args`  to disable this feature.

zsxwing commented:
> To fix this line, can we put the 600 seconds in a conf property? The other option for a RpcTimeout would look like this
> 
> BlockManagerHeartbeat(blockManagerId), new RpcTimeout(600 seconds, "BlockManagerHeartbeat")

I think this option is fine. Actually, the timeout here rarely happens because the message is sent to an Actor in the same process.

zsxwing commented:
As per the comment here: https://github.com/scala/scala/blob/2dcc4c42fdcefee08add9dbdcf619ab5da745674/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L3329

One way to prevent from this issue is removing the overload `def askWithRetry[T: ClassTag](message: Any)` and using the default parameter: `def askWithRetry[T: ClassTag](message: Any, timeout: RpcTimeout = defaultAskTimeout)`.

I just tried this locally and the compiler reported the following error:
```
[error] /Users/zsx/workspace/spark/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:189: type mismatch;
[error]  found   : scala.concurrent.duration.FiniteDuration
[error]  required: org.apache.spark.rpc.RpcTimeout
[error]       BlockManagerHeartbeat(blockManagerId), 600 seconds)
```

squito commented:
@BryanCutler good catch, I'm glad you found that!

and +1 to `-Yno-adapted-args`  (or `-Ywarn-adapted-args`) @zsxwing , I think auto-tupling is just a source of bugs

sujkh85 commented:
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-6980] [CORE] Akka timeout exceptions indicate which conf controls them (RPC Layer) (#6205)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

BryanCutler committed:
[SPARK-6980] Fixed call to askWithReply in DAGScheduler to use RpcTimeout - this was being compiled by auto-tupling and changing the message type of BlockManagerHeartbeat

BryanCutler commented:
Yeah, that would have been embarrassing if I killed the heartbeat :smile: 
I like @zsxwing 's fix, although maybe completely turning this option off in the compiler would be a good thing - is that a separate issue to discuss with more devs tho?

squito commented:
@BryanCutler yes, changing compiler options should definitely be handled as a separate issue, not as part of this PR.  sorry I should have been more clear about that, I was just agreeing the overall idea

BryanCutler committed:
[SPARK-6980] Rewrote Akka RpcTimeout UTs in RpcEnvSuite

BryanCutler committed:
[SPARK-6980] Cleaned up imports in AkkaRpcEnvSuite

BryanCutler committed:
[SPARK-6980] Removed extra import in AkkaRpcEnvSuite

BryanCutler committed:
[SPARK-6980] Fixed scala style checks

BryanCutler commented:
I rewrote the UTs from AkkaRcpEnvSuite to RpcEnvSuite and think it's ready for Jenkins to test
cc @squito

BryanCutler committed:
[SPARK-6980] Minor cleanup of test case

andrewor14 commented:
This never ran test. Jenkins, this is ok to test.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36033 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36033/consoleFull) for   PR 6205 at commit [`6a1c50d`](https://github.com/apache/spark/commit/6a1c50d4d06bc09c32f8cd9e8866070019a7e650).

SparkQA commented:
**[Test build #36033 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36033/console)**     for PR 6205 at commit [`6a1c50d`](https://github.com/apache/spark/commit/6a1c50d4d06bc09c32f8cd9e8866070019a7e650)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test PASSed.

BryanCutler committed:
[SPARK-6980] Missed one usage of deprecated RpcUtils.askTimeout in YarnSchedulerBackend although it is not being used, and fixed SparkConfSuite UT to not use deprecated RpcUtils functions

BryanCutler commented:
Hey @squito , it looks like the last Jenkins test timed out before it even got to RpcEnvSuite, can you trigger the test again?

SparkQA commented:
[Test build #992 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/992/consoleFull) for   PR 6205 at commit [`4e89c75`](https://github.com/apache/spark/commit/4e89c75f59acb8236b7c6df659a84798fe02c8ac).

squito commented on core/src/main/scala/org/apache/spark/util/RpcUtils.scala:
```diff
@@ -47,14 +47,22 @@ object RpcUtils {
   }
 
   /** Returns the default Spark timeout to use for RPC ask operations. */
+  def askRpcTimeout(conf: SparkConf): RpcTimeout = {
```
`private[spark]`

squito commented on core/src/main/scala/org/apache/spark/util/RpcUtils.scala:
```diff
@@ -47,14 +47,22 @@ object RpcUtils {
+    RpcTimeout(conf, Seq("spark.rpc.askTimeout", "spark.network.timeout"), "120s")
+  }
+
+  @deprecated("use askRpcTimeout instead", "1.5.0")
```
can you add to the message that this method was never meant to be public?

squito commented on core/src/main/scala/org/apache/spark/util/RpcUtils.scala:
```diff
@@ -47,14 +47,22 @@ object RpcUtils {
   }
 
   /** Returns the default Spark timeout to use for RPC remote endpoint lookup. */
+  def lookupRpcTimeout(conf: SparkConf): RpcTimeout = {
```
`private[spark]`

squito commented on core/src/main/scala/org/apache/spark/util/RpcUtils.scala:
```diff
@@ -47,14 +47,22 @@ object RpcUtils {
+    RpcTimeout(conf, Seq("spark.rpc.lookupTimeout", "spark.network.timeout"), "120s")
+  }
+
+  @deprecated("use lookupRpcTimeout instead", "1.5.0")
```
same here

squito commented:
Hi @BryanCutler sorry for the delay on my mind.  Thanks for updating.  Unfortunately looks like there are more conflicts with master, looks like you need to merge again.

BryanCutler commented:
No problem, I'll try to resolve these today

SparkQA commented:
**[Test build #992 timed out](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/992/console)**     for PR 6205 at commit [`4e89c75`](https://github.com/apache/spark/commit/4e89c75f59acb8236b7c6df659a84798fe02c8ac)     after a configured wait of `175m`.

BryanCutler committed:
[SPARK-6980] Changed RpcUtils askRpcTimeout and lookupRpcTimeout scope to private[spark] and improved deprecation warning msg

BryanCutler committed:
Merge branch 'master' into configTimeout-6980

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/Client.scala
	core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala
	core/src/main/scala/org/apache/spark/deploy/master/Master.scala
	core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala
	core/src/main/scala/org/apache/spark/deploy/master/ui/MasterPage.scala
	core/src/main/scala/org/apache/spark/deploy/master/ui/MasterWebUI.scala
	core/src/main/scala/org/apache/spark/deploy/rest/StandaloneRestServer.scala
	core/src/main/scala/org/apache/spark/deploy/worker/ui/WorkerPage.scala
	core/src/test/scala/org/apache/spark/rpc/akka/AkkaRpcEnvSuite.scala

BryanCutler commented:
I merged with master, which now mostly calls `askWithReply` directly.  After that I started seeing an error in running a `MasterSuite` test that complained about `RpcTimeout` not being serializable.  I changed the class definition to this
```scala
private[spark] class RpcTimeout(val duration: FiniteDuration, val timeoutProp: String)
  extends Serializable {
```
And the test was able to pass.  Does this seem correct @squito ?

squito commented:
@BryanCutler yeah making it serializable seems fine.  I have a feeling that there is probably some field that could be marked `transient` but that is not a big deal.

@zsxwing do you want to take one more pass?

squito commented:
Jenkins, retest this please

zsxwing commented:
`serializable` is fine. I forgot to add a test about sending RpcEndpointRef between different `RpcEnv`s. I will add a test in #6457.

SparkQA commented:
[Test build #995 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/995/consoleFull) for   PR 6205 at commit [`7bb70f1`](https://github.com/apache/spark/commit/7bb70f1b862099666784930ffefe92ce0fff1df3).

zsxwing commented on core/src/main/scala/org/apache/spark/util/AkkaUtils.scala:
```diff
@@ -147,7 +146,7 @@ private[spark] object AkkaUtils extends Logging {
       message: Any,
       actor: ActorRef,
-      timeout: FiniteDuration): T = {
+      timeout: RpcTimeout): T = {
```
I suggest reverting the changes to `AkkaUtils`. These methods you changed won't be called any more.

squito replied:
hmm, I disagree.  No harm in making these changes.  If we're really certain `AkkaUtils` wont' be used anymore, then we'll delete it ... but in the meantime, these changes are here just in case.

Maybe it wasn't worth the effort in the first place (my fault!) but I dont' see the harm ...

zsxwing replied:
I'm OK with the changes. Not a big deal.

zsxwing commented on core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala:
```diff
@@ -539,6 +544,97 @@ abstract class RpcEnvSuite extends SparkFunSuite with BeforeAndAfterAll {
+          context.reply(msg)
+        }
+        case sr: SleepyReply => {
+          Thread.sleep(50)
```
`Thread.sleep(50)` may cause this test becomes a flaky one. How about:
```
    val sleepLatch = new CountDownLatch(1)

    val rpcEndpointRef = env.setupEndpoint("ask-future", new RpcEndpoint {
      override val rpcEnv = env

      override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
        case msg: String => {
          context.reply(msg)
        }
        case sr: SleepyReply => {
          sleepLatch.await()
          context.reply(sr.msg)
        }
      }
    })

    val longTimeout = new RpcTimeout(1 second, "spark.rpc.long.timeout")
    val shortTimeout = new RpcTimeout(10 millis, "spark.rpc.short.timeout")

    // Ask with immediate response, should complete successfully
    val fut1 = rpcEndpointRef.ask[String]("hello", longTimeout)
    val reply1 = longTimeout.awaitResult(fut1)
    assert("hello" === reply1)

    // Ask with a delayed response and wait for response immediately that should timeout
    val fut2 = rpcEndpointRef.ask[String](SleepyReply("doh"), shortTimeout)
    val reply2 =
      intercept[RpcTimeoutException] {
        shortTimeout.awaitResult(fut2)
      }.getMessage

    // RpcTimeout.awaitResult should have added the property to the TimeoutException message
    assert(reply2.contains(shortTimeout.timeoutProp))

    // Ask with delayed response and allow the Future to timeout before Await.result
    val fut3 = rpcEndpointRef.ask[String](SleepyReply("goodbye"), shortTimeout)

    // Allow future to complete with failure using plain Await.result, this will return
    // once the future is complete to verify addMessageIfTimeout was invoked
    val reply3 =
      intercept[RpcTimeoutException] {
        Await.result(fut3, 200 millis)
      }.getMessage

    sleepLatch.countDown()
```

squito replied:
+1, great idea.

maybe is it also worth adding one more check that `SleepyReply` does make it through after you release the latch?  Or if that isn't necessary, couldn't we just change `SleepyReply` to `NeverReply`, which never sends anything back and we don't even need the latch.

BryanCutler replied:
Good ideas, thanks!  For the purpose of this test, it is not necessary to get a reply, so `NeverReply` works pretty well.  I'll push this change

zsxwing commented:
LGTM except two minor comments.

BryanCutler committed:
[SPARK-6980] RpcTimeout class extends Serializable, was causing error in MasterSuite

BryanCutler committed:
[SPARK-6980] Changed RpcEnvSuite test to never reply instead of just sleeping, to avoid possible sync issues

SparkQA commented:
[Test build #995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/995/console) for   PR 6205 at commit [`7bb70f1`](https://github.com/apache/spark/commit/7bb70f1b862099666784930ffefe92ce0fff1df3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait ExpectsInputTypes `
  * `trait AutoCastInputTypes `

SparkQA commented:
[Test build #996 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/996/consoleFull) for   PR 6205 at commit [`46c8d48`](https://github.com/apache/spark/commit/46c8d48c7eb17d668daac7aca080f5e5bfce18a6).

SparkQA commented:
[Test build #996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/996/console) for   PR 6205 at commit [`46c8d48`](https://github.com/apache/spark/commit/46c8d48c7eb17d668daac7aca080f5e5bfce18a6).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait ExpectsInputTypes `
  * `trait AutoCastInputTypes `

SparkQA commented:
[Test build #997 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/997/consoleFull) for   PR 6205 at commit [`46c8d48`](https://github.com/apache/spark/commit/46c8d48c7eb17d668daac7aca080f5e5bfce18a6).

SparkQA commented:
[Test build #997 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/997/console) for   PR 6205 at commit [`46c8d48`](https://github.com/apache/spark/commit/46c8d48c7eb17d668daac7aca080f5e5bfce18a6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class Heartbeat(workerId: String, worker: RpcEndpointRef) extends DeployMessage`
  * `  case class RegisteredWorker(master: RpcEndpointRef, masterWebUiUrl: String) extends DeployMessage`
  * `  case class RegisterApplication(appDescription: ApplicationDescription, driver: RpcEndpointRef)`
  * `  case class RegisteredApplication(appId: String, master: RpcEndpointRef) extends DeployMessage`
  * `  case class SubmitDriverResponse(`
  * `  case class KillDriverResponse(`
  * `  case class MasterChanged(master: RpcEndpointRef, masterWebUiUrl: String)`
  * `class DCT(override val uid: String)`
  * `class MinMaxScaler(override val uid: String)`
  * `class PCA (override val uid: String) extends Estimator[PCAModel] with PCAParams `
  * `class StreamingLinearAlgorithm(object):`
  * `class StreamingLinearRegressionWithSGD(StreamingLinearAlgorithm):`
  * `class AnalysisException(Exception):`
  * `case class Cast(child: Expression, dataType: DataType) extends UnaryExpression with Logging `
  * `trait ExpectsInputTypes `
  * `trait AutoCastInputTypes `
  * `trait ExtractValue `
  * `abstract class ExtractValueWithStruct extends UnaryExpression with ExtractValue `
  * `abstract class ExtractValueWithOrdinal extends BinaryExpression with ExtractValue `
  * `      class SpecificOrdering extends $`
  * `    class SpecificProjection extends $`
  * `    final class SpecificRow extends $`
  * `case class Crc32(child: Expression)`
  * `    // compiled class file for the closure here will conflict with the one in callUDF (upper case).`

squito commented:
merged to master.

Thanks for all your hard work on this @BryanCutler!  this was probably a lot more work than you initially expected (and me too), but I'm glad you stuck with it, its a great addition.

BryanCutler commented:
Awesome, thanks for all your help @squito!  That was a heck of a first Jira, but it was a great experience and I definitely learned a lot.
